### PR TITLE
RING-44448 export_s3_keys.sh improvements

### PR DESCRIPTION
**Pull request by TrevorBenson**
RING-44448 - keep raw files option
RING-44448 - determine the IP to query for bucketd
RING-44448 - basic error logs for each RID and bucket

I'm opening this as a draft PR for us to discuss and for some initial testing. 

1. It asks about raw files up front and keeps them all or none.
2. Uses `ss` to determine if port 9000 is bound to a specific IP or not. 
   * Matches `[::]:9000`for when ipv6 is enabled, or disabled via sysctl.
   * Matches `0.0.0.0:9000` when ipv6 is disabled via grub and `[::]` will not appear in `ss` output.
3. There is a very basic log output for INFO, WARNING and ERROR. 
   * A basic RAW_COUNT vs. PROCESSED_COUNT is performed.
   * Empty buckets are logged as WARNINGS.

Currently the script will report any failures that occurs into the `RID_${RID}.log` as well as the `${bucket}.log` file, then it will exit with a specific RC code as defined in the script. This "fail early" logic will stop the processing of the RAFT session and report back which buckets have not yet been processed at the time of the ERROR and exiting the script.

### CODE REVIEW

**Comment by TrevorBenson** [comment]
Currently, the script has a hardcoded s3utils version which has to be bumped when the offline archive bundles a newer version. 

Options:
1. Manage the script by a github action. When the s3utils version is bumped, the `export_s3_keys.sh` would get a the new versioned hardcoded.
 2. Manage `export_s3_keys.sh` as a Jinja template. Would make the script deployed/updated by Ansible, potentially having an older version if Ansible is not used.
* Check container hosts local registry for all s3utils versions. Determine the highest one, and use it.

Something like:
```
vercomp () {
    if [[ $1 == $2 ]]; then
        return 0
    fi
    local IFS=.
    local i
    local ver1=($1)
    local ver2=($2)
    for ((i=${#ver1[@]}; i<${#ver2[@]}; i++)); do
        ver1[i]=0
    done
    for ((i=0; i<${#ver1[@]}; i++)); do
        if [[ -z ${ver2[i]} ]]; then
            ver2[i]=0
        fi
        if ((10#${ver1[i]} > 10#${ver2[i]})); then
            return 1
        fi
        if ((10#${ver1[i]} < 10#${ver2[i]})); then
            return 2
        fi
    done
    return 0
}


S3UTILS_VERSION=($(podman images registry.scality.com/s3utils/s3utils --format '{{ .Tag }}'))
echo found versions ${S3UTILS_VERSION[*]}
while [[ ${#S3UTILS_VERSION[@]} -gt 1 ]]; do
    element1=${S3UTILS_VERSION[0]}
    for ((i=1; i<${#S3UTILS_VERSION[@]}; i++)); do
        element2=${S3UTILS_VERSION[i]}
        vercomp $element1 $element2
        case $? in
            1) S3UTILS_VERSION=("${S3UTILS_VERSION[@]:0:$i}" "${S3UTILS_VERSION[@]:$i+1}"); i=$((i-1));;
            2) S3UTILS_VERSION=("${S3UTILS_VERSION[@]:1:$i}" "${S3UTILS_VERSION[@]:$i+1}"); i=$((i-1));;
        esac
    done
done
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 1]
```diff
@@ -1,32 +1,191 @@
...
+UNBOUND_BUCKETD=('[::]:9000' '0.0.0.0:9000')
+WORKDIR=/var/tmp/bucketSproxydKeys
+
+# Check if a raft id was passed
+if [[ -z ${1} ]]; then
+	echo "Usage: ${0} <raft_id>"
+	exit "${RC_NO_RAFT_ID}"
+else
+	RID=${1}
+	echo -e "\nWARNING: Don't forget to check for adequate disk space."
```
Comment (scality-fno) [inline scripts/S3_FSCK/export_s3_keys.sh:25]:
```suggestion
	echo -e "\nWARNING: Don't forget to check for adequate disk space in ${WORKDIR}."
```

**Comment by TrevorBenson** [comment]
@scality-fno @fra-scality any additional suggestions before we widen the pool of reviewers?

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 2]
```diff
@@ -1,32 +1,191 @@
...
+		else
+			echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Found bucketd instance ${BUCKETD_HOST}" | tee -a "${WORKDIR}/RID_${RID}.log"
+		fi
+	fi
+}
+
+# shellcheck disable=SC2312
+function list_buckets_of_rid() {
+	RAFT_OUTPUT=$(curl --silent "http://${BUCKETD_HOST}:9000/_/raft_sessions/${RID}/bucket")
+	mapfile -t BUCKETS < <(echo "${RAFT_OUTPUT}" | jq -r '.[] | select (. | contains("mpuShadowBucket") | not) | select (. | contains("users..bucket") | not)')
```
Comment (scality-fno) [inline scripts/S3_FSCK/export_s3_keys.sh:83]:
```suggestion
	mapfile -t BUCKETS < <(echo "${RAFT_OUTPUT}" | jq -r '.[] | select((contains("mpuShadowBucket")|not) and (contains("users..bucket")|not))')
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 3]
```diff
@@ -1,32 +1,191 @@
...
+		stat=$?
+	fi
+	return "${stat}"
+}
+
+# shellcheck disable=SC2312
+function init() {
+	if ! [[ -d ${WORKDIR} ]]; then
+		if ! mkdir -pv "${WORKDIR}"; then
+			echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") ERROR: Failed to create ${WORKDIR}"
```
Comment (scality-fno) [inline scripts/S3_FSCK/export_s3_keys.sh:53]:
```suggestion
			echo "$(date -u +"%FT%TZ") ERROR: Failed to create ${WORKDIR}"
```

**Review by scality-fno — COMMENTED** [review]
haven't been able to test this one... do you have a lab handy with everything in place from this "dev" environment?
That said, gave a few suggestions (pointless, essentially).
Question: is the double tee pipeline used for logging somewhat redundant?

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 4]
```diff
@@ -1,32 +1,191 @@
...
+			exit "${RC_FILE_PERMISSIONS}"
+		else
+			echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Created ${WORKDIR}" | tee -a "${WORKDIR}/RID_${RID}.log"
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:56]:
```suggestion
			echo "$(date -u +"%FT%TZ") INFO: Created ${WORKDIR}" | tee -a "${WORKDIR}/RID_${RID}.log"
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 5]
```diff
@@ -1,32 +1,191 @@
...
+	if [[ ${SS_OUTPUT} == "${UNBOUND_BUCKETD[0]}" ]] || [[ ${SS_OUTPUT} == "${UNBOUND_BUCKETD[1]}" ]]; then
+		BUCKETD_HOST="127.0.0.1"
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Found bucketd instance matched UNBOUND_BUCKET. Using BUCKETD_HOST: ${BUCKETD_HOST}" | tee -a "${WORKDIR}/RID_${RID}.log"
+	else
+		BUCKETD_HOST=$(echo "${SS_OUTPUT}" | cut -d: -f1)
+		if [[ -z ${BUCKETD_HOST} ]] || [[ $(is_valid_ip "${BUCKETD_HOST}") == 0 ]]; then
+			echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") ERROR: Failed to find bucketd instance" | tee -a "${WORKDIR}/RID_${RID}.log"
+			exit "${RC_NO_BUCKETD_FOUND}"
+		else
+			echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Found bucketd instance ${BUCKETD_HOST}" | tee -a "${WORKDIR}/RID_${RID}.log"
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:75]:
```suggestion
		echo "$(date -u +"%FT%TZ") INFO: Found bucketd instance matched UNBOUND_BUCKET. Using BUCKETD_HOST: ${BUCKETD_HOST}" | tee -a "${WORKDIR}/RID_${RID}.log"
	else
		BUCKETD_HOST=$(echo "${SS_OUTPUT}" | cut -d: -f1)
		if [[ -z ${BUCKETD_HOST} ]] || [[ $(is_valid_ip "${BUCKETD_HOST}") == 0 ]]; then
			echo "$(date -u +"%FT%TZ") ERROR: Failed to find bucketd instance" | tee -a "${WORKDIR}/RID_${RID}.log"
			exit "${RC_NO_BUCKETD_FOUND}"
		else
			echo "$(date -u +"%FT%TZ") INFO: Found bucketd instance ${BUCKETD_HOST}" | tee -a "${WORKDIR}/RID_${RID}.log"
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 6]
```diff
@@ -1,32 +1,191 @@
...
+
+# shellcheck disable=SC2312
+function list_buckets_of_rid() {
+	RAFT_OUTPUT=$(curl --silent "http://${BUCKETD_HOST}:9000/_/raft_sessions/${RID}/bucket")
+	mapfile -t BUCKETS < <(echo "${RAFT_OUTPUT}" | jq -r '.[] | select (. | contains("mpuShadowBucket") | not) | select (. | contains("users..bucket") | not)')
+	if [[ ${#BUCKETS[@]} -eq 0 ]]; then
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") WARNING: No buckets found in raft session ${RID}" | tee -a "${WORKDIR}/RID_${RID}.log"
+		exit "${RC_NO_BUCKETS_FOUND_IN_RID}"
+	fi
+	echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Found ${#BUCKETS[@]} buckets in raft session ${RID}" | tee -a "${WORKDIR}/RID_${RID}.log"
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:88]:
```suggestion
		echo "$(date -u +"%FT%TZ") WARNING: No buckets found in raft session ${RID}" | tee -a "${WORKDIR}/RID_${RID}.log"
		exit "${RC_NO_BUCKETS_FOUND_IN_RID}"
	fi
	echo "$(date -u +"%FT%TZ") INFO: Found ${#BUCKETS[@]} buckets in raft session ${RID}" | tee -a "${WORKDIR}/RID_${RID}.log"
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 7]
```diff
@@ -1,32 +1,191 @@
...
+		--entrypoint /usr/local/bin/node \
+		-e 'BUCKETD_HOSTPORT=127.0.0.1:9000' \
+		-e "BUCKETS=${bucket}" \
+		-e 'NO_MISSING_KEY_CHECK=1' \
+		-e 'VERBOSE=1' \
+		registry.scality.com/s3utils/s3utils:1.14.0 \
+		verifyBucketSproxydKeys.js \
+		>"${WORKDIR}"/raw_"${bucket}"_keys.txt; then
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") ERROR: Failed to export keys for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") ERROR: Unprocessed buckets: ${UNPROCESSED_BUCKETS[*]}" | tee -a "${WORKDIR}/RID_${RID}".log
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:107]:
```suggestion
		echo "$(date -u +"%FT%TZ") ERROR: Failed to export keys for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
		echo "$(date -u +"%FT%TZ") ERROR: Unprocessed buckets: ${UNPROCESSED_BUCKETS[*]}" | tee -a "${WORKDIR}/RID_${RID}".log
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 8]
```diff
@@ -1,32 +1,191 @@
...
+
+# shellcheck disable=SC2312
+function check_for_completed_scan() {
+	# See if the raw file contains the message "completed scan"
+	SCAN_COMPLETE=$(jq -r '. | select(.message | contains("completed scan"))' "${WORKDIR}/raw_${bucket}_keys.txt" | jq -s length)
+	if [[ ${SCAN_COMPLETE} -eq 0 ]]; then
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") WARNING: verifyBucketSproxydKeys scan did not complete for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
+		exit "${RC_EXPORT_S3_KEYS_FAILURE}"
+	else
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: verifyBucketSproxydKeys scan completed for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:120]:
```suggestion
		echo "$(date -u +"%FT%TZ") WARNING: verifyBucketSproxydKeys scan did not complete for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
		exit "${RC_EXPORT_S3_KEYS_FAILURE}"
	else
		echo "$(date -u +"%FT%TZ") INFO: verifyBucketSproxydKeys scan completed for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 9]
```diff
@@ -1,32 +1,191 @@
...
+	fi
+}
+
+# shellcheck disable=SC2312
+function process_bucket_contents() {
+	echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Processing output of RID ${RID} - bucket ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
+	if ! jq -r '. | select(.message | contains("sproxyd key"))  + {"bucket": .objectUrl  } | .bucket |= sub("s3://(?<bname>.*)/.*"; .bname) | .objectUrl |= sub("s3://.*/(?<oname>.*)$"; .oname) | [.bucket, .objectUrl, .sproxydKey] | @csv' "${WORKDIR}/raw_${bucket}_keys.txt" >"${WORKDIR}/${bucket}_keys.txt"; then
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") ERROR: Failed to process keys for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") ERROR: Unprocessed buckets: ${UNPROCESSED_BUCKETS[*]}" | tee -a "${WORKDIR}/RID_${RID}".log
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:129]:
```suggestion
	echo "$(date -u +"%FT%TZ") INFO: Processing output of RID ${RID} - bucket ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
	if ! jq -r '. | select(.message | contains("sproxyd key"))  + {"bucket": .objectUrl  } | .bucket |= sub("s3://(?<bname>.*)/.*"; .bname) | .objectUrl |= sub("s3://.*/(?<oname>.*)$"; .oname) | [.bucket, .objectUrl, .sproxydKey] | @csv' "${WORKDIR}/raw_${bucket}_keys.txt" >"${WORKDIR}/${bucket}_keys.txt"; then
		echo "$(date -u +"%FT%TZ") ERROR: Failed to process keys for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
		echo "$(date -u +"%FT%TZ") ERROR: Unprocessed buckets: ${UNPROCESSED_BUCKETS[*]}" | tee -a "${WORKDIR}/RID_${RID}".log
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 10]
```diff
@@ -1,32 +1,191 @@
...
+	PROCESSED_COUNT=$(wc -l <"${WORKDIR}/${bucket}_keys.txt")
+	if [[ ${RAW_COUNT} -ne ${PROCESSED_COUNT} ]]; then
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") WARNING: Raw and processed counts do not match for ${bucket} - RAW_COUNT: ${RAW_COUNT} != PROCESED_COUNT: ${PROCESSED_COUNT}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") ERROR: Unprocessed buckets: ${UNPROCESSED_BUCKETS[*]}" | tee -a "${WORKDIR}/RID_${RID}".log
+		exit "${RC_COUNT_MISMATCH}"
+	else
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Raw and processed counts match for ${bucket} - RAW_COUNT: ${RAW_COUNT} == PROCESED_COUNT: ${PROCESSED_COUNT}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
+		# if the raw and processed counts match but are both 0 then print a warning that the bucket is empty
+		if [[ ${RAW_COUNT} -eq 0 ]]; then
+			echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") WARNING: Bucket ${bucket} appears to be empty: RAW_COUNT: ${RAW_COUNT} == PROCESED_COUNT: ${PROCESSED_COUNT}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:146]:
```suggestion
		echo "$(date -u +"%FT%TZ") WARNING: Raw and processed counts do not match for ${bucket} - RAW_COUNT: ${RAW_COUNT} != PROCESED_COUNT: ${PROCESSED_COUNT}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
		echo "$(date -u +"%FT%TZ") ERROR: Unprocessed buckets: ${UNPROCESSED_BUCKETS[*]}" | tee -a "${WORKDIR}/RID_${RID}".log
		exit "${RC_COUNT_MISMATCH}"
	else
		echo "$(date -u +"%FT%TZ") INFO: Raw and processed counts match for ${bucket} - RAW_COUNT: ${RAW_COUNT} == PROCESED_COUNT: ${PROCESSED_COUNT}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
		# if the raw and processed counts match but are both 0 then print a warning that the bucket is empty
		if [[ ${RAW_COUNT} -eq 0 ]]; then
			echo "$(date -u +"%FT%TZ") WARNING: Bucket ${bucket} appears to be empty: RAW_COUNT: ${RAW_COUNT} == PROCESED_COUNT: ${PROCESSED_COUNT}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 11]
```diff
@@ -1,32 +1,191 @@
...
+	fi
+
+}
+
+# shellcheck disable=SC2312
+function cleanup() {
+	if [[ ${KEEP_RAW_FILES} =~ ^[Yy]$ ]]; then
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Keeping raw files for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
+	else
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Removing raw files for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:157]:
```suggestion
		echo "$(date -u +"%FT%TZ") INFO: Keeping raw files for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
	else
		echo "$(date -u +"%FT%TZ") INFO: Removing raw files for ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 12]
```diff
@@ -1,32 +1,191 @@
...
+		rm -f "${WORKDIR}/raw_${bucket}_keys.txt"
+	fi
+}
+
+# shellcheck disable=SC2312
+function export_buckets() {
+	for bucket in "${BUCKETS[@]}"; do
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Starting export of ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:165]:
```suggestion
		echo "$(date -u +"%FT%TZ") INFO: Starting export of ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```

File: scripts/S3_FSCK/export_s3_keys.sh [hunk 13]
```diff
@@ -1,32 +1,191 @@
...
+		# Export the bucket contents from the Raft Session
+		export_bucket_contents
+		# Check for scan completed
+		check_for_completed_scan
+		# Process the raw file into a csv
+		process_bucket_contents
+		# Check the raw and processed counts match
+		compare_raw_and_processed_counts
+		echo
+		echo "$(date -u +"%Y-%m-%dT%H:%M:%SZ") INFO: Completed export of ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```
Comment (TrevorBenson) [inline scripts/S3_FSCK/export_s3_keys.sh:175]:
```suggestion
		echo "$(date -u +"%FT%TZ") INFO: Completed export of ${bucket}" | tee -a "${WORKDIR}/RID_${RID}".log | tee -a "${WORKDIR}/${bucket}".log
```

**Comment by TrevorBenson** [comment]
> haven't been able to test this one... do you have a lab handy with everything in place from this "dev" environment? That said, gave a few suggestions (pointless, essentially). Question: is the double tee pipeline used for logging somewhat redundant?

Sort of, but not exactly. There are items that are only in the RID logs because they are not specific to a single bucket. THe RID log however does contain every log entry from each bucket. Below is the breakdown of errors for a bucket log, and then the additional log entries the RID log will have.

## TL;DR

* The `<bucket>.log` contains
  * Everything for its own bucket, including:
    * INFO about Starting/Completing export of bucket
    * INFO (or) WARNING about **verifyBucketSproxydKeys.js** scan completion
    * ERROR about failed to export keys or process keys
    * INFO/WARNING about raw and processed count comparisons
    * INFO about keep/remove of raw files

* The `PID_<PID>.log` contains
  * Everything inside each `<bucket>.log`
  * Only present in the PID logs:
    * INFO/ERROR about finding bucketd instance (valid IP etc.)
    * INFO about the list of buckets per Raft ID
    * ERROR about Raft ID empty (ENOENT / no buckets found)
    * ERROR list of all unprocessed buckets whenever dumping the entire RID does not end successfully

**Comment by TrevorBenson** [comment]
Opening this up to a wider audience.
